server/src/main/java/org/elasticsearch/common/xcontent/support/XContentMapValues.java

```diff
@@ -195,7 +195,18 @@ private static Object extractValue(String[] pathElements, int index, Object curr
             for (Object o : valueList) {
                 Object listValue = extractValue(pathElements, index, o, nullValue);
                 if (listValue != null) {
-                    newList.add(listValue);
+                    // we add arrays as list elements only if we are already at leaf,
+                    // otherwise append individual elements to the new list so we don't
+                    // accumulate intermediate array structures
+                    if (listValue instanceof List<?> list) {
+                        if (index == pathElements.length) {
+                            newList.add(list);
+                        } else {
+                            newList.addAll(list);
+                        }
+                    } else {
+                        newList.add(listValue);
+                    }
```

Comment:
Can we add something to `XContentMapValuesTests` that exercises this?

Reply:
Sure, I didn't see that particular test class but it makes much sense to do more in-depth testing there then.